### PR TITLE
Feature/gui hooks

### PR DESCRIPTION
**Pull request by dinhlongviolin1**
Related to Avaiga/taipy-designer#126

### CODE REVIEW

File: taipy/gui/hooks.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        self.__hooks.append(hook)
+
+    def __getattr__(self, name: str):
+        def _resolve_hook(*args, **kwargs):
+            for hook in self.__hooks:
+                if name not in hook.exposed_hooks:
+                    continue
+                # call hook
+                try:
+                    res = getattr(hook, name)(*args, **kwargs)
```
Comment (FredLL-Avaiga) [inline taipy/gui/hooks.py]: Check that this attribute is a function?

**Review by FredLL-Avaiga — DISMISSED** [review]
Great idea
Could we cache the first valid answer to speed things?

File: taipy/gui/hooks.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import typing as t
+
+from taipy.logger._taipy_logger import _TaipyLogger
+
+from .utils.singleton import _Singleton
+
+
+class BaseGuiHook:
```
Comment (FabienLelaquais) [inline taipy/gui/hooks.py]:
```suggestion
class GuiHook:
```

File: taipy/gui/hooks.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    exposed_hooks: t.List[str] = []
```
Comment (FabienLelaquais) [inline taipy/gui/hooks.py]:
```suggestion
    hooks: t.List[str] = []
```

File: taipy/gui/hooks.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class GuiHooks(object, metaclass=_Singleton):
+    def __init__(self):
+        self.__hooks: t.List[BaseGuiHook] = []
```
Comment (FabienLelaquais) [inline taipy/gui/hooks.py]:
```suggestion
        self.__hooks: t.List[GuiHook] = []
```

File: taipy/gui/hooks.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    def _register_gui_hook(self, hook: BaseGuiHook):
```
Comment (FabienLelaquais) [inline taipy/gui/hooks.py]:
```suggestion
    def _register_gui_hook(self, hook: GuiHook):
```

File: taipy/gui/hooks.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        self.__hooks.append(hook)
+
+    def __getattr__(self, name: str):
+        def _resolve_hook(*args, **kwargs):
+            for hook in self.__hooks:
+                if name not in hook.exposed_hooks:
```
Comment (FabienLelaquais) [inline taipy/gui/hooks.py]:
```suggestion
                if name not in hook.hooks:
```

File: taipy/gui/hooks.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                    continue
+                # call hook
+                try:
+                    res = getattr(hook, name)(*args, **kwargs)
+                except Exception as e:
+                    _TaipyLogger._get_logger().error(f"Error while calling Gui Hooks '{name}': {e}")
```
Comment (FabienLelaquais) [inline taipy/gui/hooks.py]:
```suggestion
                    _TaipyLogger._get_logger().error(f"Error while calling Gui hook '{name}': {e}")
```

**Review by FabienLelaquais — DISMISSED** [review]
Great!
Refactoring proposed...

File: taipy/gui/hooks.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+                    return
+                # check if the hook returns True -> stop the chain
+                if res == True:
```
Comment (FredLL-Avaiga) [inline taipy/gui/hooks.py]: use is True to check the type as well

File: taipy/gui/hooks.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+                    return
+                # only hooks that return true are allowed to return values to ensure consistent response
+                if isinstance(res, (list, tuple)) and len(res) == 2 and res[0] == True:
```
Comment (FredLL-Avaiga) [inline taipy/gui/hooks.py]: is True

File: taipy/gui/hooks.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+                    return res[1] or None
```
Comment (FredLL-Avaiga) [inline taipy/gui/hooks.py]: why the `or None` ? that would change the returned value if res[2] is False or "" or empty iterable

File: taipy/gui/hooks.py (hunk 1 above)
Comment (dinhlongviolin1) [reply taipy/gui/hooks.py]: That is an excellent point! thanks!!!

File: taipy/gui/hook.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import typing as t
+
+from taipy.logger._taipy_logger import _TaipyLogger
+
+from .utils.singleton import _Singleton
+
+
+class Hook:
+    method_names: t.List[str] = []
```
Comment (FabienLelaquais) [inline taipy/gui/hook.py:9]: Is this line useful?

**Review by FabienLelaquais — APPROVED** [review]
👍

File: taipy/gui/hook.py (hunk 11 above)
Comment (dinhlongviolin1) [reply taipy/gui/hook.py:9]: it is, very. As a sense of base class implementation

**Review by FredLL-Avaiga — APPROVED** [review]
Great
Not interested in caching ?

**Review by namnguyen20999 — APPROVED** [review]
lgtm

**Pull request merged**